specs/http-client2/index.html

```diff
@@ -255,14 +285,23 @@ <h3>Request Options</h3>
         </p>
 
         <dl>
-          <dt><code>status-only</code> (<code>xs:boolean</code>)</dt>
-          <dd>Skip body in response. Default: false.</dd>
+          <dt><code>http-version</code> (<code>xs:string</code>)</dt>
+          <dd>The HTTP Version to use for the request. Default: 1.1. Supported versions are <code>1.0</code> and <code>1.1</code>. An unsupported version raises the error: TODO-ERROR-CODE.</dd>
+
+          <dt><code>headers</code> (<code>map(xs:string, xs:string)</code>)</dt>
+          <dd>HTTP header fields.</dd>
+
+          <dt><code>content-encoding</code> (<code>xs:string</code>)</dt>
```

Comment:
Isn’t `content-encoding` a usual header field, similar to `content-type` (I might be wrong)?

Reply:
So yes, `Content-Encoding` is a HTTP Header. However using such a Content Encoding requires that the content be encoded ;-)

So for example, rather than force the user to GZip encode the content themselves, we can help them out by easily doing that encoding for them; I wanted to add some pre-baked options for the two most widely used encodings to make it easier for the users.

How about if I added the text:

    Applies the specified encoding to body, and sets the HTTP Header Content-Encoding`, this overrides any Contet-Encoding header specified in $options.

Reply:
If we decide to allow `Content-Encoding` to be specified one level above the `headers` entry, I would go further:

1. We could define a set of the most popular header fields to be specified on top. It would also make some of my examples more compact if `Content-Type` doesn’t need to be embedded in the `headers` field.
2. We could send all options as headers fields which are no predefined HTTP Client options.

And, additionally, we should probably raise an error if a field is supplied both on top and in the `headers` entry.

Reply:
Ah! That isn't really what I am trying to achieve here at all. I think perhaps I was not very clear at explaining it. Let me try again:

I considered `content-encoding` *option* in the `$options` map to be special because it requires the HTTP Client implementation to actually change how it sends the data to the server. i.e. this `content-encoding` *option* is controlling how the HTTP Client module processes the body of the HTTP Request; It is **not only** about the HTTP Header!
Setting the `content-encoding` *option* in the `$options` causes two things to happen:
1. The `Content-Encoding` HTTP Header will be explicitly set in the HTTP request
2. The body of the HTTP Request will be auto-encoded according to the `content-encoding` *option* , i.e. it may be gzip'd or deflated before it is sent to the http server.

If you don't like the use of `content-encoding` as a key in `$options`, we could use something else. I could of instead added two key's `"gzip": false()` and `"deflate": false()` however as they are mutually exclusive, and both describe content encoding actions, I thought it better just to call them `content-encoding`.

To be clear, the `content-encoding` *option* is not the same thing as the `Content-Encoding` HTTP Header ;-)

---

Secondly, I realise that I have done something inconsistent in this PR already :-/
Consider the `chunked` *option*. It is similar to the `content-encoding` *option*,  i.e. it does two things:

1. The `Transfer-Encoding` HTTP Header will be explicitly set in the HTTP request
2. The HTTP request body will be sent to the http server in a chunked format automatically by the HTTP Client.

If you are happy with the explanation of the `content-encoding` *option* being a different thing than the `Content-Encoding` header, I think that it is likely that the `chunked` *option* should actually be renamed as the `transfer-encoding` *option*.

Hope that all makes sense?

Reply:
Thanks for the clarification! I now see what this is about.

I would have actually guessed that an implementation of the HTTP Client Module will automatically gzip data if `gzip` is supplied as value of the `Content-Encoding` field – similar to responses, which we can – and should – automatically unzip if the `Content-Encoding` response header is set to `gzip`. Would that make sense to you? I think the same applies to the `Transfer-Encoding` field.

Otherwise, we’d have to deal with cases in which a custom `Content-Encoding` or `Transfer-Encoding` header field is assigned by a user (raise an error; interpret the top-level option, but don’t overwrite the header field; …).

Reply:
I had a bit of a different take on it. II think we have to have some rules around implicit and explicit serialization for both the request and the response. I have yet to get to that part of the spec and my thinking on it yet ;-)

I had rather envisaged here that the user can set any http headers they want (which includes the Content-Encoding and Transfer-Coding headers), and that we take no implicit action on them, they supply the content in the body and we just sent it as is (ignoring any other Content-Type implicit/explicit serialization rules). However, if they specify a content-type *option* or `transfer-encoding` *option*, then we do *special stuff* with the body and override the appropriate headings.

I should probably further read your proposals about request response handling and then come back to this.

Reply:
True: In the current proposal, the `Content-Type` header isn’t triggering any conversions; instead, the media-type will be assigned to that field if no other value has been assigned to it by the user. So it’s probably consistent if we don’t introduce any other fields with implicit behavior.

To avoid confusion, though, I like the idea of having names that clearly differ. Both `chunked` and `gzip` would sound good to me.